lib/IRGen/TypeInfo.h

```diff
@@ -477,6 +477,14 @@ class TypeInfo {
                                    Address src, llvm::Value *count,
                                    SILType T) const;
 
+  /// Outlining helper function: recursively traverse the SILType:
+  /// When encountering an Archetype - add it to a type-metadata vec.
+  virtual void collectArchetypeMetadata(
```

Comment:
You can implement this without iterating over TypeInfo by walking the SILType instead, and then you won't need to override it in each subsclass, etc.